src/lib/core/data-table/data-table.ts

```diff
@@ -0,0 +1,189 @@
+ */
+@Component({
+  selector: 'cdk-table',
+  template: `
+    <ng-container cdkHeaderRowPlaceholder></ng-container>
+    <ng-container cdkRowPlaceholder></ng-container>
+  `,
+  host: {
+    'class': 'cdk-table',
+    'role': 'grid' // TODO(andrewseguin): Allow the user to choose either grid or treegrid
+  },
+  encapsulation: ViewEncapsulation.None,
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class CdkTable {
+  /**
+   * Provides a stream containing the latest data array to render. Influenced by the table's
+   * stream of view window (what rows are currently on screen).
+   */
+  @Input() dataSource: DataSource<any>;
+
+  /**
+   * Stream containing the latest information on what rows are being displayed on screen.
+   * Can be used by the data source to as a heuristic of what data should be provided.
+   */
+  // TODO(andrewseguin): Remove max value as the end index
+  // and instead calculate the view on init and scroll.
+  viewChanges = new BehaviorSubject<CollectionViewer>({start: 0, end: Number.MAX_VALUE});
+
+  /**
+   * Map of all the user's defined columns identified by name.
+   * Contains the header and data-cell templates.
+   */
+  private _columnDefinitionsByName = new Map<string,  CdkColumnDef>();
+
+  // Placeholders within the table's template where the header and data rows will be inserted.
+  @ViewChild(CdkRowPlaceholder) _rowPlaceholder: CdkRowPlaceholder;
+  @ViewChild(CdkHeaderRowPlaceholder) _headerRowPlaceholder: CdkHeaderRowPlaceholder;
+
+  /**
+   * The column definitions provided by the user that contain what the header and cells should
+   * render for each column.
+   */
+  @ContentChildren(CdkColumnDef) _columnDefinitions: QueryList<CdkColumnDef>;
+
+  /** Template used as the header container. */
+  @ContentChild(CdkHeaderRowDef) _headerDefinition: CdkHeaderRowDef;
+
+  /** Set of templates that used as the data row containers. */
+  @ContentChildren(CdkRowDef) _rowDefinitions: QueryList<CdkRowDef>;
+
+  constructor(private _changeDetectorRef: ChangeDetectorRef) {}
+
+  ngOnDestroy() {
+    // TODO(andrewseguin): Disconnect from the data source so
+    // that it can unsubscribe from its streams.
+  }
+
+  ngOnInit() {
+    // TODO(andrewseguin): Setup a listener for scroll events
+    // and emit the calculated view to this.viewChanges
+  }
+
+  ngAfterContentInit() {
+    // TODO(andrewseguin): Throw an error if two columns share the same name
+    this._columnDefinitions.forEach(columnDef => {
+      this._columnDefinitionsByName.set(columnDef.name, columnDef);
+    });
+  }
+
+  ngAfterViewInit() {
+    // TODO(andrewseguin): Re-render the header when the header's columns change.
+    this.renderHeaderRow();
+
+    // TODO(andrewseguin): Re-render rows when their list of columns change.
+    // TODO(andrewseguin): If the data source is not
+    // present after view init, connect it when it is defined.
+    this.dataSource.connectTable(this.viewChanges).subscribe((rowsData: any[]) => {
```

Comment:
This looks like a promising solution to subscriptions across a lot of our components. I'll add a TODO to resolve this. Likely if we go this route, we'll want to put it in several other components as well so it might be best in another PR